tekton-pipelines/README.md

```diff
@@ -0,0 +1,216 @@
+  ```
+- List all taskruns:
+  ```sh
+  kubectl get taskruns
+  # or
+  tkn taskrun list
+  ```
+
+## Create and Run a Pipeline
+- Create a pipeline:
+  ```sh
+  kubectl apply -f pipelines/link-pull-request-to-jira.yaml
+  ```
+- List all pipelines:
+  ```
+  kubectl get pipelines
+  # or
+  tkn pipeline list
+  ```
+- Run a pipeline:
+  Manually create a PipelineRun named `link-pull-request-to-jira-run-1`:
+  ```yaml
+  apiVersion: tekton.dev/v1beta1
+  kind: PipelineRun
+  metadata:
+    name: link-pull-request-to-jira-run-1
+  spec:
+    params:
+    - name: PR_NUMBER
+      value: "1"
+    - name: REPO_FULL_NAME
+      value: vfreex/elliott
+    pipelineRef:
+      name: link-pull-request-to-jira
+  ```
+  Or use the `tkn` command:
+  ```sh
+  # interactively
+  tkn pipeline start link-pull-request-to-jira
+  # or non-interactively
+  tkn pipeline start link-pull-request-to-jira -p REPO_FULL_NAME=vfreex/elliott -p PR_NUMBER=1
+  ```
+- See logs:
+  ``` sh
+  tkn pipelinerun logs --last -f
+  ```
+- List all pipelines:
+  ```sh
+  kubectl get pipelineruns
+  # or
+  tkn pipelinerun list
+  ```
+
+## Automatically Trigger a PipelineRun using GitHub Webhooks
+- Add a TriggerBinding to convert GitHub webhook payload to parameters and a TriggerTemplate to template PipelineRun resource:
+  ```sh
+  kubectl apply -f ./triggers/link-pull-request-to-jira.yaml
+  ```
+- Add an EventListener.
+
+  Edit `secrets/github-webhook.yaml` for your webhook secret then add it to Kubernetes:
+  ```sh
+  kubectl apply -f secrets/github-webhook.yaml
+  ```
+  Create the EventListener:
+  ```sh
+  kubectl apply -f infra/rbac.yaml
+  kubectl apply -f infra/github-event-listener.yaml
+  ```
+- (On a public cloud) Expose the EventListener service via Ingress or Route.
+  ``` sh
+  # OpenShift
+  oc expose service/el-github-event-listener
+  ```
+- Add a webhook to your GitHub repo:
+  ```
+  # Settings -> Webhook -> Add webhook
+  Payload URL: your exposed event listener URL
+  Content type: application/json
+  Secret: your secret in secrets/github-webhook.yaml
+  Which events would you like to trigger this webhook? "Let me select individual events." Check "Pull requests".
+  Active: check
+  ```
+- Test your trigger
+  Forward your local 8080 port to the event listener pod:
+  ```sh
+  kubectl get po -l eventlistener=github-event-listener
+  kubectl port-forward <pod-name> 8080
+  ```
+  Send a webhook payload. For example:
+  ```sh
+  curl -v \
+    -H 'X-GitHub-Event: pull_request' \
+    -H 'X-Hub-Signature: sha1=13eaa0168f8d8efcdf5189ea75b782cf89809de6' \
+    -H 'Content-Type: application/json' \
+    -d '{"action": "opened", "head_commit":{"id":"master"},"repository":{"url": "https://github.com/tektoncd/triggers"}}' \
+    http://localhost:8080
+  ```
+  See [Webhook event payloads][] for the format.
+## (Optional) Proxy GitHub webhook behind a secure firewall with smee.io
+- Create a new channel at https://smee.io/.
+- Run `smee` CLI behind your firewall.
```

Comment:
i think we just need the smee client running in a container and handling all our external events generically. i already wanted this for art-bot. listening for github webhooks is an obvious use case. i think we can just provide a configmap for mapping known endpoints to internal targets. if this gets serious enough we could look into [hosting our own smee.io](https://github.com/probot/smee.io#deploying-your-own-smeeio) on say api.ci... if DPTP hasn't already set up a similar relay. smee.io is public and so is this repo, so we'd need to be careful about our traffic since basically anyone could figure out how to watch it and spoof it. we'll need to think that through at some point.

Reply:
Right, smee.io can be watched and spoofed and it doesn't provide any obvious security measures by itself. 

In this demo we are relying on the GitHub interceptor https://github.com/openshift/aos-cd-jobs/pull/2335/files#diff-7e0d17996c28fa4a4296278fb9565a8aR10-R16 to verify if the webhook request is actually sent by GitHub (by checking the signature generated from the secret defined at https://github.com/openshift/aos-cd-jobs/pull/2335/files#diff-b8176e4e814062ec7824d9ab9b698c9aR7). The content is still visible by anyone viewing smee.io but I am not sure if that part should also be protected...

Reply:
i don't think we care about the webhook being observable as long as it can't be spoofed. so that seems fine; i wasn't aware of that detail, thanks.